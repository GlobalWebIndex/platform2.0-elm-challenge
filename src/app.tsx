import { FC, PropsWithChildren } from 'react';
import { BrowserRouter } from 'react-router-dom';
import ModalProvider from 'mui-modal-provider';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

const queryClient = new QueryClient();

export const App: FC<PropsWithChildren> = ({ children }) => {
  return (
    <BrowserRouter>
      <QueryClientProvider client={queryClient}>
        <ModalProvider>{children}</ModalProvider>
      </QueryClientProvider>
    </BrowserRouter>
  );
};
