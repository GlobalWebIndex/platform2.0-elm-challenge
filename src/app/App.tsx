import React, { useEffect } from 'react';
import { Box, Flex } from '@chakra-ui/react';
import {
  Redirect,
  Route,
  Switch,
} from 'react-router-dom';
import { Header, NoMatch } from '../common/components';
import { FavouritesPage, fetchFavourites } from '../favourites';
import { SingleBreedPage, BreedsPage } from '../breeds';
import { useAppDispatch } from '../store';
import { ImagesPage } from '../images';

export const App = (): JSX.Element => {
  const dispatch = useAppDispatch();

  // Prefetching
  useEffect(() => {
    const isNotInFavouritesPage = window.location.pathname !== '/favourites';

    if (isNotInFavouritesPage) {
      dispatch(fetchFavourites());
    }
  }, []);

  return (
    <Flex direction="column" h="100vh">
      <Header />
      <Box h="100%" p={5}>
        <Switch>
          <Route path="/images">
            <ImagesPage />
          </Route>
          <Route path="/breeds/:breedId">
            <SingleBreedPage />
          </Route>
          <Route path="/breeds">
            <BreedsPage />
          </Route>
          <Route exact path="/favourites">
            <FavouritesPage />
          </Route>
          <Route exact path="/">
            <Redirect to="/images" />
          </Route>
          <Route path="*">
            <NoMatch />
          </Route>
        </Switch>
      </Box>
    </Flex>
  );
};
