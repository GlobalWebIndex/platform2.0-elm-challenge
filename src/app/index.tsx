/**
 *
 * App
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 */

import * as React from 'react';
import { Helmet } from 'react-helmet-async';
import { Switch, Route } from 'react-router-dom';

import { GlobalStyle } from 'styles/global-styles';

import { NotFoundPage } from './components/NotFoundPage/Loadable';
import { useTranslation } from 'react-i18next';
import { NavigationItems } from './routes';

export function App() {
  const { i18n } = useTranslation();
  return (
    <>
      <Helmet
        titleTemplate="%s - Cat App"
        defaultTitle="React App"
        htmlAttributes={{ lang: i18n.language }}
      >
        <meta name="description" content="A cat lover's application" />
      </Helmet>
      <Switch>
        {NavigationItems.map(item => (
          <Route
            exact
            path={item.route}
            key={item.name}
            component={item.component}
          />
        ))}
        <Route component={NotFoundPage} />
      </Switch>

      <GlobalStyle />
    </>
  );
}
