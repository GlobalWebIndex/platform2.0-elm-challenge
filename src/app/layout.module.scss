.container {
  height: 100vh;
  width: 100vw;
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: 372px 1fr 100px;
  gap: 30px;
  grid-template-areas:
    "header"
    "main"
    "footer";

  & > main {
    margin: 0 10%;
    display: flex;
    flex-direction: column;
    align-items: stretch;
    gap: 30px;
    grid-area: main;
  }

  & > header {
    grid-area: header;
  }

  & > footer {
    grid-area: footer;
    height: 100px;
  }
}
