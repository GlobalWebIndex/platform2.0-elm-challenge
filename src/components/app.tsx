import React from 'react';
import {Navigate, Route, Routes, useLocation} from 'react-router';
import {BreedList} from './breedList';
import {CatDetailsPreview} from './catDetailsPreview';
import {CatList} from './catList';
import {Layout} from './layout';
import {Modal} from './modal';

export const App = (): JSX.Element => {
  const location = useLocation();
  const state = location.state as {backgroundLocation?: string} | null;

  return (
    <>
      <Routes location={state?.backgroundLocation ?? location}>
        <Route element={<Layout />} path="/">
          <Route element={<Navigate replace to="cats" />} index />
          <Route element={<CatList />} path="cats" />
          <Route element={<CatDetailsPreview />} path="cats/:id" />
          <Route element={<BreedList />} path="breeds">
            <Route element={<Modal />} path=":id" />
          </Route>
          <Route element={<Navigate replace to="cats" />} path="*" />
        </Route>
      </Routes>
      {state?.backgroundLocation && (
        <Routes>
          <Route element={<Modal />} path="cats/:id" />
        </Routes>
      )}
    </>
  );
};
