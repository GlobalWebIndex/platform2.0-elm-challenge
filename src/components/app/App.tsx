import React from 'react';
import { BrowserRouter, Route, Link, Switch } from "react-router-dom";
import { Navbar, Button, Alignment, Classes } from "@blueprintjs/core"

import "@blueprintjs/core/lib/css/blueprint.css";
import "@blueprintjs/icons/lib/css/blueprint-icons.css";

/* Components */
import Breeds from "../breeds/Breeds"
import Cats from "../cats/Cats"
import Favorites from '../favorites/Favorites'
import Page404 from '../error-pages/Page404';
import { ScreenClassProvider } from 'react-grid-system';
import { ModalProvider } from 'react-modal-hook';


const App: React.FC = () => {

  return (
    <React.Fragment>
      <BrowserRouter>
        <ModalProvider>
          <ScreenClassProvider>
            <Navbar className="bp3-dark">
              <Navbar.Group align={Alignment.LEFT}>
                <Navbar.Heading>GWI Cats Challenge</Navbar.Heading>
                <Navbar.Divider />
                <Link to="/">
                  <Button className={Classes.MINIMAL} icon="random" text="Browse Images" />
                </Link>
                <Link to="/breeds">
                  <Button className={Classes.MINIMAL} icon="code" text="Breeds" />
                </Link>
                <Link to="/favorites">
                  <Button className={Classes.MINIMAL} icon="heart" text="Favorites" />
                </Link>
              </Navbar.Group>
            </Navbar>
            {/* Routes */}
            <Switch>
              <Route path="/" exact component={Cats} />
              <Route path="/cats/:image_id" component={Cats} />
              <Route path="/breeds" component={Breeds} />
              <Route path="/favorites" component={Favorites} />
              <Route path="/*" component={Page404} status={404} />
            </Switch>
          </ScreenClassProvider>
        </ModalProvider>
      </BrowserRouter>
    </React.Fragment>
  );
}

export default App;
