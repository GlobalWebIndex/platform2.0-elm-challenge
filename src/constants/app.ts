export const BASE_URL = 'https://api.thecatapi.com/v1/';
export const API_KEY = '7d7e1e32-8db8-4919-863c-aa44e502627b';
export const USER_ID = '7d7e1e32-8db8-4919-863c-aa44e502627b';
export const PAGINATION_DEFAULT_PAGE = 0;
