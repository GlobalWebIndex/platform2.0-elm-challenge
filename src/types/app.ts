export type DefaultStateType = {
  loading: boolean;
  hasError: boolean;
  error: string;
};

export type DefaultPaginationStateType<T> = DefaultStateType & {
  page: number;
  list: T[];
};

export type ErrorPayload = {
  error: string;
};
