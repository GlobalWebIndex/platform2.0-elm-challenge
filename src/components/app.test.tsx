import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import {render} from '@testing-library/react';
import {App} from './app';

jest.mock('./breedList', () => ({
  BreedList: () => {
    const {Outlet} = jest.requireActual('react-router');

    return (
      <div>
        Breed list
        <Outlet />
      </div>
    );
  }
}));

jest.mock('./catDetailsPreview', () => ({
  CatDetailsPreview: () => <div>Cat details preview</div>
}));

jest.mock('./catList', () => ({
  CatList: () => <div>Cat list</div>
}));

jest.mock('./modal', () => ({
  Modal: () => <div>Modal</div>
}));

describe('<App />', () => {
  it('should render the index route "cats"', () => {
    const {container} = render(
      <MemoryRouter initialEntries={['/']}>
        <App />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  describe('when the route is "/cats"', () => {
    it('should render the cat list', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/cats']}>
          <App />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('when the route is "/cats/:id"', () => {
    it('should render the cat details preview', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/cats/1']}>
          <App />
        </MemoryRouter>
      );
      expect(container).toMatchSnapshot();
    });

    describe('when a background location exists', () => {
      it('should render the modal', () => {
        const {container} = render(
          <MemoryRouter initialEntries={[{pathname: '/cats/1', state: {backgroundLocation: '/cats'}}]}>
            <App />
          </MemoryRouter>
        );
        expect(container).toMatchSnapshot();
      });
    });
  });

  describe('when the route is "/breeds"', () => {
    it('should render the breed list', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/breeds']}>
          <App />
        </MemoryRouter>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('when the route is "/breeds/:id"', () => {
    it('should render the modal', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/breeds/1']}>
          <App />
        </MemoryRouter>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('when the route is unknown', () => {
    it('should redirect to the "/cats" route', () => {
      const {container} = render(
        <MemoryRouter initialEntries={['/unknown']}>
          <App />
        </MemoryRouter>
      );
      expect(container).toMatchSnapshot();
    });
  });
});
