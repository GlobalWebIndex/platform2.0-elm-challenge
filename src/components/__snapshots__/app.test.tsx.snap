// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render the index route "cats" 1`] = `
<div>
  <div
    class="navigation"
  >
    <h1>
      Cat lover
    </h1>
    <nav>
      <ul>
        <li>
          <a
            href="/cats"
          >
            Cats
          </a>
        </li>
        <li>
          <a
            href="/breeds"
          >
            Breeds
          </a>
        </li>
      </ul>
    </nav>
  </div>
  <div>
    Cat list
  </div>
</div>
`;

exports[`<App /> when the route is "/breeds" should render the breed list 1`] = `
<div>
  <div
    class="navigation"
  >
    <h1>
      Cat lover
    </h1>
    <nav>
      <ul>
        <li>
          <a
            href="/cats"
          >
            Cats
          </a>
        </li>
        <li>
          <a
            href="/breeds"
          >
            Breeds
          </a>
        </li>
      </ul>
    </nav>
  </div>
  <div>
    Breed list
  </div>
</div>
`;

exports[`<App /> when the route is "/breeds/:id" should render the modal 1`] = `
<div>
  <div
    class="navigation"
  >
    <h1>
      Cat lover
    </h1>
    <nav>
      <ul>
        <li>
          <a
            href="/cats"
          >
            Cats
          </a>
        </li>
        <li>
          <a
            href="/breeds"
          >
            Breeds
          </a>
        </li>
      </ul>
    </nav>
  </div>
  <div>
    Breed list
    <div>
      Modal
    </div>
  </div>
</div>
`;

exports[`<App /> when the route is "/cats" should render the cat list 1`] = `
<div>
  <div
    class="navigation"
  >
    <h1>
      Cat lover
    </h1>
    <nav>
      <ul>
        <li>
          <a
            href="/cats"
          >
            Cats
          </a>
        </li>
        <li>
          <a
            href="/breeds"
          >
            Breeds
          </a>
        </li>
      </ul>
    </nav>
  </div>
  <div>
    Cat list
  </div>
</div>
`;

exports[`<App /> when the route is "/cats/:id" should render the cat details preview 1`] = `
<div>
  <div
    class="navigation"
  >
    <h1>
      Cat lover
    </h1>
    <nav>
      <ul>
        <li>
          <a
            href="/cats"
          >
            Cats
          </a>
        </li>
        <li>
          <a
            href="/breeds"
          >
            Breeds
          </a>
        </li>
      </ul>
    </nav>
  </div>
  <div>
    Cat details preview
  </div>
</div>
`;

exports[`<App /> when the route is "/cats/:id" when a background location exists should render the modal 1`] = `
<div>
  <div
    class="navigation"
  >
    <h1>
      Cat lover
    </h1>
    <nav>
      <ul>
        <li>
          <a
            href="/cats"
          >
            Cats
          </a>
        </li>
        <li>
          <a
            href="/breeds"
          >
            Breeds
          </a>
        </li>
      </ul>
    </nav>
  </div>
  <div>
    Cat list
  </div>
  <div>
    Modal
  </div>
</div>
`;

exports[`<App /> when the route is unknown should redirect to the "/cats" route 1`] = `
<div>
  <div
    class="navigation"
  >
    <h1>
      Cat lover
    </h1>
    <nav>
      <ul>
        <li>
          <a
            href="/cats"
          >
            Cats
          </a>
        </li>
        <li>
          <a
            href="/breeds"
          >
            Breeds
          </a>
        </li>
      </ul>
    </nav>
  </div>
  <div>
    Cat list
  </div>
</div>
`;
